private/protocol/decode.go

```diff
@@ -0,0 +1,112 @@
+package protocol
```

Comment:
probably don't want to commit the decode.go file for now as this isn't really supported by the SDK and design is needed for the decode logic.  We can stash this and associated decode files in a feature branch until we're ready to use them.